Let us start with the code for using Naive Bayes method to classify truthful vs. deceptive hotel reviews.

*Importing Libraries*


In [15]:
import pandas as pd
import numpy as np

*Reading the files and doing the only data preprocessing we used, applying lowercase to all words*


In [16]:
def Reading(fname1, fname2):
    file=open(fname1)
    
    true=file.readlines()
    
    for i in range(len(true)):
        
        true[i]=true[i].lower()
        
    file=open(fname2)
    
    fake=file.readlines()
    
    for i in range(len(fake)):
        
        fake[i]=fake[i].lower()
        
    return(true,fake)


*A function to combine the truthful and fake reviews into dataframes with the appropriate labels, for both training and validation sets*

In [17]:

def Combining(list1,list2):

    df_reviews_train_t=pd.DataFrame()

    df_reviews_train_t=pd.DataFrame(list1)

    df_reviews_train_t.columns={'Reviews'}

    df_reviews_train_t['Label']=0

   

    df_reviews_train_f=pd.DataFrame()

    df_reviews_train_f=pd.DataFrame(list2)

    df_reviews_train_f.columns={'Reviews'}

    df_reviews_train_f['Label']=1

   

    Final=[df_reviews_train_t,df_reviews_train_f ]

    result=pd.concat(Final)

    return(result)


In [18]:
#Training Set
Reviews_train_t,Reviews_train_f=Reading('truthful.txt','deceptive.txt')

DF_Reviews_train=Combining(Reviews_train_t,Reviews_train_f)

#Validation Set
Reviews_val_t,Reviews_val_f=Reading('truthfulval.txt','deceptiveval.txt')

DF_Reviews_val=Combining(Reviews_val_t,Reviews_val_f)

*Function to read test file and apply lowecase on it.*

In [19]:
def Reading_test(fname1):

    file=open(fname1)

    test=file.readlines()

    for i in range(len(test)):

        test[i]=test[i].lower()

    return(test)

In [21]:
Reviews_test=Reading_test('test.txt')

Let us import the packages and methods required for Naive Bayes Implementation. We'll be trying out two type of NB, Multinomial and Bernoulli.


In [22]:
from sklearn.feature_extraction.text import CountVectorizer

from sklearn.naive_bayes import MultinomialNB

from sklearn.naive_bayes import BernoulliNB

from sklearn.metrics import accuracy_score

Count Vectorizer

In [23]:
#Fit training data
vect = CountVectorizer().fit(DF_Reviews_train['Reviews'])

#Transform training, validation and test sets

X_train_vectorized = vect.transform(DF_Reviews_train['Reviews'])
X_Val_vectorized = vect.transform(DF_Reviews_val['Reviews'])
test_vectorized= vect.transform(Reviews_test)

Let us try both types of Naive Bayes on the Validation set and see how it goes

In [24]:
#Multinomial 

#Alpha = 1

clfrNB = MultinomialNB(alpha = 1)

clfrNB.fit(X_train_vectorized,DF_Reviews_train['Label'] )

preds = clfrNB.predict(vect.transform(DF_Reviews_val['Reviews']))

 
score =  accuracy_score(DF_Reviews_val['Label'], preds)

print('Multinomial Accuracy on Validation with Alpha 1: '+ str(score*100))

#Alpha = 0.1

clfr1NB = MultinomialNB(alpha = 0.1)

clfr1NB.fit(X_train_vectorized,DF_Reviews_train['Label'] )

preds = clfr1NB.predict(vect.transform(DF_Reviews_val['Reviews']))

 
score =  accuracy_score(DF_Reviews_val['Label'], preds)

print('Multinomial Accuracy on Validation with Alpha 0.1: '+ str(score*100))

Multinomial Accuracy on Validation with Alpha 1: 91.40625
Multinomial Accuracy on Validation with Alpha 0.1: 90.234375


In [25]:
#Bernoulli

#Alpha = 1

BerNB = BernoulliNB(alpha = 1)

BerNB.fit(X_train_vectorized,DF_Reviews_train['Label'] )

preds = BerNB.predict(vect.transform(DF_Reviews_val['Reviews']))

 
score =  accuracy_score(DF_Reviews_val['Label'], preds)

print('Bernoulli Accuracy on Validation with Alpha 1: '+ str(score*100))

#Alpha = 0.1

BernNB = BernoulliNB(alpha = 0.1)

BernNB.fit(X_train_vectorized,DF_Reviews_train['Label'] )

preds = BernNB.predict(vect.transform(DF_Reviews_val['Reviews']))

 
score =  accuracy_score(DF_Reviews_val['Label'], preds)

print('Bernoulli Accuracy on Validation with Alpha 0.1: '+ str(score*100))

Bernoulli Accuracy on Validation with Alpha 1: 87.890625
Bernoulli Accuracy on Validation with Alpha 0.1: 88.671875


Preparing submission file from both Naive Bayes Classifiers

In [26]:
#Alpha = 0.1
test_preds_Bern = BernNB.predict(test_vectorized)
#Aplha = 1
test_preds_Bern_1 = BerNB.predict(test_vectorized)
#Alpha = 1
test_preds_Multi_1 = clfrNB.predict(test_vectorized)
#Alpha = 0.1
test_preds_Multi = clfr1NB.predict(test_vectorized)                                

After converting them to CSV's, we found out that Bernoulli NB(Alpha = 0.1) had the highest accuracy on Kaggle of 91.406%. 

In [27]:
print(test_preds_Bern)

[0 0 0 1 0 0 1 1 1 0 1 1 1 0 1 1 1 0 1 0 0 0 1 1 1 1 1 1 1 0 1 1 0 0 1 0 1
 0 1 1 0 1 1 1 1 1 0 1 1 1 0 0 0 1 1 1 0 0 0 1 1 1 0 0 0 1 0 0 0 1 0 1 0 0
 0 1 1 1 1 1 0 0 0 0 0 0 0 1 0 0 1 0 0 0 1 1 1 1 0 0 0 0 0 0 0 0 0 1 1 1 0
 1 0 1 1 0 0 1 1 1 0 0 1 0 0 1 0 1 0 1 0 0 1 1 0 0 0 1 0 1 0 1 0 0 0 0 0 1
 0 0 0 1 1 1 1 1 1 1 1 1 0 0 1 0 0 1 1 1 1 0 0 0 0 1 0 1 1 0 0 0 1 1 0 1 1
 1 0 0 1 1 1 0 1 1 0 0 0 1 1 1 0 1 1 1 1 1 0 0 1 1 0 1 1 0 0 1 1 1 0 1 0 0
 0 0 0 0 0 1 0 1 0 1 0 0 0 1 1 1 0 0 1 0 1 1 0 1 1 1 1 0 0 1 1 1 0 1 1 0 0
 0 0 0 1 0 1 0 1 0 0 0 1 1 0 0 1 1 1 1 0 1 0 0 1 1 0 1 0 0 1 1 0 0 1 1 0 1
 0 0 0 0 1 1 0 1 1 0 0 1 1 1 1 0 1 0 1 1 1 1 0 1]


Given below in the layout we used to convert them to CSV's.

In [ ]:
output = pd.DataFrame({'Id': test_data.index,
                       'Prediction': test_preds})
                       
                       
                       
output.to_csv('submission_test.csv', index=False)